File: ios-native/ClubFuoco/Features/Events/EventCards.swift
import SwiftUI

// Event cards in the explore feed's "cinema" style — the same geometry, type
// and materials as HeroCard and LandCard next to them (220pt hero photo with
// the info panel below on Theme.surface; 220×130 landscape card with overlay
// text), so an event reads as one more card in the feed rather than a
// transplant from a different design.
//
// What differs is the CONTENT, not the styling. An event leads with its own
// name and who is playing, and names the venue underneath; a venue card leads
// with the venue. Everything here goes through Theme tokens, so both cards
// follow light and dark mode with the rest of the feed.

private struct EventPhoto: View {
    let url: String?
    let height: CGFloat
    /// Largest display dimension (points), so the photo is fetched right-sized
    /// instead of at its stored 800px. nil = fetched native.
    var targetWidth: CGFloat? = nil

    var body: some View {
        Theme.imagePlaceholder
            .overlay {
                if let url, let parsed = URL(string: url) {
                    CachedAsyncImage(url: parsed, targetWidth: targetWidth) { image in
                        image.resizable().aspectRatio(contentMode: .fill)
                    } placeholder: {
                        Theme.imagePlaceholder
                    }
                } else {
                    // Not the venue glyph the club cards fall back to — an
                    // event without a flyer is still an event, not a room.
                    Image(systemName: "sparkles")
                        .font(.system(size: 28))
                        .foregroundStyle(Theme.fadedSand.opacity(0.4))
                }
            }
            .frame(height: height)
            .clipped()
    }
}

/// The corner marker, in the feed's filled-pill idiom (see TagPill on the venue
/// cards). Only ever ONE of these, and the order below is the priority:
/// tonight is time-critical, a pin is our editorial choice, "ours" is
/// provenance. Stacking all three would turn the card into a badge shelf.
///
/// The promoter's paid `featured` flag is deliberately NOT surfaced as a badge
/// — the buyer gets rank in the feed, not a label telling guests they paid.
private struct EventTag: View {
    let event: FeedEvent
    @Environment(LocaleStore.self) private var locale

    var body: some View {
        if event.isTonight {
            pill(locale.t("events.tonightTag"), background: Theme.ember, color: .white)
        } else if event.pinned {
            // Dark ink on gold, not white: white on #C09950 is about 2:1 and
            // unreadable at 9pt.
            pill(locale.t("events.pickTag"), background: Theme.gold, color: Color(hex: 0x221E1A))
        } else if event.house {
            pill(locale.t("events.oursTag"), background: .black.opacity(0.5), color: .white.opacity(0.85))
        }
    }

    private func pill(_ text: String, background: Color, color: Color) -> some View {
        Text(text.uppercased())
            .font(.cfSans(9, weight: .medium))
            .kerning(1)
            .foregroundStyle(color)
            .padding(.horizontal, 8)
            .padding(.vertical, 3)
            .background(background, in: .capsule)
    }
}

// ── Hero (the pinned event, above the venue shelves) ──────────────────────────

struct EventHeroCard: View {
    let event: FeedEvent
    @Environment(LocaleStore.self) private var locale

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            ZStack(alignment: .topLeading) {
                EventPhoto(url: event.image, height: 220)
                    .overlay(
                        LinearGradient(
                            stops: [
                                .init(color: .black.opacity(0.45), location: 0),
                                .init(color: .clear, location: 0.5),
                                .init(color: .black.opacity(0.3), location: 1),
                            ],
                            startPoint: .top, endPoint: .bottom
                        )
                    )

                HStack(alignment: .top) {
                    EventTag(event: event)
                    Spacer()
                }
                .padding(12)

                if event.isFree {
                    VStack {
                        Spacer()
                        HStack {
                            Spacer()
                            Text(locale.t("events.free").uppercased())
                                .font(.cfSans(10, weight: .semibold))
                                .kerning(0.8)
                                .foregroundStyle(.white)
                                .padding(.horizontal, 9)
                                .padding(.vertical, 4)
                                .background(.black.opacity(0.55), in: .capsule)
                        }
                    }
                    .padding(12)
                }
            }
            .frame(height: 220)

            VStack(alignment: .leading, spacing: 6) {
                Text(locale.t("events.kicker").uppercased())
                    .font(.cfSans(9))
                    .kerning(1.3)
                    .foregroundStyle(Theme.fadedSand)

                Text(event.displayTitle)
                    .font(.cfSerif(30, italic: true))
                    .foregroundStyle(Theme.accent)
                    .lineLimit(2)
                    .minimumScaleFactor(0.8)

                // The line-up takes the slot the venue card gives its address,
                // hairline and all. It is the billing, and people choose a
                // night by who is playing; the blurb only shows when there are
                // no credits to bill.
                if let secondary = event.lineupLine() ?? event.description?.nilIfBlank {
                    Text(secondary.prefix(90))
                        .font(.cfSerif(13, italic: true))
                        .foregroundStyle(Theme.stone)
                        .lineLimit(2)
                        .padding(.leading, 10)
                        .overlay(alignment: .leading) {
                            Rectangle().fill(Theme.hairline).frame(width: 2)
                        }
                }

                HStack {
                    Text(event.metaLine(locale: locale))
                        .font(.cfSans(11))
                        .foregroundStyle(Theme.fadedSand)
                        .lineLimit(1)
                    Spacer(minLength: 8)
                    HStack(spacing: 6) {
                        Text(locale.t("explore.viewEvent"))
                            .font(.cfSans(13, weight: .semibold))
                            .fixedSize()
                        Image(systemName: "arrow.right")
                            .font(.system(size: 11, weight: .semibold))
                    }
                    .foregroundStyle(Theme.cream)
                    .padding(.horizontal, 16)
                    .padding(.vertical, 9)
                    .background(Theme.ink, in: .capsule)
                }
                .padding(.top, 6)
            }
            .padding(.init(top: 16, leading: 20, bottom: 18, trailing: 20))
        }
        .background(Theme.surface)
        .clipShape(.rect(cornerRadius: 16))
        .shadow(color: Color(hex: 0x221E1A).opacity(0.06), radius: 12, y: 8)
    }
}

// ── Rail card (220×130, overlay text — LandCard's geometry) ───────────────────

struct EventCard: View {
    let event: FeedEvent
    @Environment(LocaleStore.self) private var locale

    var body: some View {
        ZStack(alignment: .bottomLeading) {
            EventPhoto(url: event.image, height: 130, targetWidth: FeedImage.thumbWidth)
                .overlay(
                    LinearGradient(
                        stops: [
                            .init(color: .black.opacity(0.7), location: 0),
                            .init(color: .clear, location: 0.55),
                        ],
                        startPoint: .bottom, endPoint: .top
                    )
                )

            VStack(alignment: .leading, spacing: 1) {
                Text(event.displayTitle)
                    .font(.cfSans(13, weight: .semibold))
                    .foregroundStyle(.white)
                    .lineLimit(1)
                // Day and place only. The full meta line carries door times
                // too, which truncates away at 220pt — the times are on the
                // event page, one tap down. `placeLine` is the whole route on a
                // night that moves, not just the venue it starts at.
                Text([event.dayLabel(locale: locale), event.placeLine]
                    .compactMap { $0 }.joined(separator: " · "))
                    .font(.cfSans(10))
                    .foregroundStyle(.white.opacity(0.6))
                    .lineLimit(1)
            }
            .padding(.horizontal, 10)
            .padding(.bottom, 8)

            VStack {
                HStack(alignment: .top) {
                    EventTag(event: event)
                    Spacer()
                }
                Spacer()
            }
            .padding(7)
        }
        .frame(width: 220, height: 130)
        .clipShape(.rect(cornerRadius: 12))
        .shadow(color: Color(hex: 0x221E1A).opacity(0.06), radius: 8, y: 4)
    }
}

private extension String {
    var nilIfBlank: String? {
        trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ? nil : self
    }
}

File: android-native/app/src/main/java/com/clubfuoco/app/features/explore/EventCards.kt
package com.clubfuoco.app.features.explore

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowForward
import androidx.compose.material.icons.filled.AutoAwesome
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.font.FontStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.clubfuoco.app.R
import com.clubfuoco.app.core.designsystem.FeedImage
import com.clubfuoco.app.core.designsystem.FuocoImage
import com.clubfuoco.app.core.designsystem.Geist
import com.clubfuoco.app.core.designsystem.InstrumentSerif
import com.clubfuoco.app.core.designsystem.Theme
import com.clubfuoco.app.core.designsystem.clickableUnlessBusy
import com.clubfuoco.app.models.FeedEvent

/**
 * Event cards in the explore feed's "cinema" style — the same geometry, type and
 * materials as [HeroCard] and [PosterCard] next to them, so an event reads as
 * one more card in the feed rather than a transplant from another design.
 *
 * What differs is the CONTENT, not the styling. An event leads with its own name
 * and who is playing, and names the venue underneath; a venue card leads with
 * the venue.
 *
 * Port of `EventCards.swift`.
 */

/**
 * The corner marker. Only ever ONE, and the order here is the priority: tonight
 * is time-critical, a pin is our editorial choice, "ours" is provenance.
 * Stacking all three would turn the card into a badge shelf.
 *
 * The promoter's paid `featured` flag is deliberately NOT a badge — the buyer
 * gets rank in the feed, not a label telling guests they paid.
 */
@Composable
private fun EventTag(event: FeedEvent) {
    when {
        event.isTonight -> TagPill(
            stringResource(R.string.events_tonightTag).uppercase(),
            Theme.accent,
            Color.White,
        )
        // Dark ink on gold, not white: white on #C09950 is about 2:1 and
        // unreadable at this size.
        event.pinned -> TagPill(
            stringResource(R.string.events_pickTag).uppercase(),
            Theme.gold,
            Color(0xFF221E1A),
        )
        event.house -> TagPill(
            stringResource(R.string.events_oursTag).uppercase(),
            Color.Black.copy(alpha = 0.5f),
            Color.White.copy(alpha = 0.85f),
        )
    }
}

@Composable
private fun EventPhoto(url: String?, height: androidx.compose.ui.unit.Dp, targetWidth: androidx.compose.ui.unit.Dp? = null) {
    Box(
        Modifier.fillMaxWidth().height(height).background(Theme.surface),
        contentAlignment = Alignment.Center,
    ) {
        if (url != null) {
            if (targetWidth != null) {
                FuocoImage(url, Modifier.fillMaxSize(), targetWidth = targetWidth)
            } else {
                FuocoImage(url, Modifier.fillMaxSize())
            }
        } else {
            // Not the venue glyph the club cards fall back to — an event without
            // a flyer is still an event, not a room.
            Icon(
                Icons.Filled.AutoAwesome,
                contentDescription = null,
                tint = Theme.fadedSand.copy(alpha = 0.4f),
                modifier = Modifier.size(28.dp),
            )
        }
    }
}

// ── Hero (the pinned event, leading the featured box) ────────────────────────

@Composable
fun EventHeroCard(event: FeedEvent, tonight: String, tomorrow: String, onOpen: () -> Unit) {
    Column(
        Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(16.dp))
            .background(Theme.surface)
            .clickableUnlessBusy(onClick = onOpen),
    ) {
        Box(Modifier.fillMaxWidth().height(220.dp)) {
            EventPhoto(event.image, 220.dp)
            Box(
                Modifier.fillMaxSize().background(
                    Brush.verticalGradient(
                        0.0f to Color.Black.copy(alpha = 0.45f),
                        0.5f to Color.Transparent,
                        1.0f to Color.Black.copy(alpha = 0.3f),
                    ),
                ),
            )
            Box(Modifier.align(Alignment.TopStart).padding(12.dp)) { EventTag(event) }
            if (event.isFree) {
                Box(Modifier.align(Alignment.BottomEnd).padding(12.dp)) {
                    TagPill(
                        stringResource(R.string.events_free).uppercase(),
                        Color.Black.copy(alpha = 0.55f),
                        Color.White,
                    )
                }
            }
        }

        Column(
            Modifier.padding(start = 20.dp, end = 20.dp, top = 16.dp, bottom = 18.dp),
            verticalArrangement = Arrangement.spacedBy(6.dp),
        ) {
            Text(
                stringResource(R.string.events_kicker).uppercase(),
                fontFamily = Geist, fontSize = 9.sp, letterSpacing = 1.3.sp,
                color = Theme.fadedSand,
            )
            Text(
                event.displayTitle,
                fontFamily = InstrumentSerif, fontStyle = FontStyle.Italic,
                fontSize = 30.sp, color = Theme.accent,
                maxLines = 2, overflow = TextOverflow.Ellipsis,
            )

            // The line-up takes the slot the venue card gives its address,
            // hairline and all. It is the billing, and people choose a night by
            // who is playing; the blurb only shows when there is nothing to bill.
            val secondary = event.lineupLine()
                ?: event.description?.trim()?.takeIf { it.isNotEmpty() }
            secondary?.let {
                Row(horizontalArrangement = Arrangement.spacedBy(10.dp)) {
                    Box(Modifier.width(2.dp).height(34.dp).background(Theme.hairline))
                    Text(
                        it.take(90),
                        fontFamily = InstrumentSerif, fontStyle = FontStyle.Italic,
                        fontSize = 13.sp, color = Theme.stone,
                        maxLines = 2, overflow = TextOverflow.Ellipsis,
                    )
                }
            }

            Row(
                Modifier.fillMaxWidth().padding(top = 6.dp),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Text(
                    event.metaLine(tonight, tomorrow),
                    fontFamily = Geist, fontSize = 11.sp, color = Theme.fadedSand,
                    maxLines = 1, overflow = TextOverflow.Ellipsis,
                    modifier = Modifier.weight(1f),
                )
                Spacer(Modifier.width(8.dp))
                Row(
                    Modifier
                        .clip(CircleShape)
                        .background(Theme.ink)
                        .padding(horizontal = 16.dp, vertical = 9.dp),
                    verticalAlignment = Alignment.CenterVertically,
                    horizontalArrangement = Arrangement.spacedBy(6.dp),
                ) {
                    Text(
                        stringResource(R.string.explore_viewEvent),
                        fontFamily = Geist, fontWeight = FontWeight.SemiBold, fontSize = 13.sp,
                        color = Theme.cream,
                    )
                    Icon(
                        Icons.AutoMirrored.Filled.ArrowForward,
                        contentDescription = null,
                        tint = Theme.cream,
                        modifier = Modifier.size(12.dp),
                    )
                }
            }
        }
    }
}

// ── Rail card ────────────────────────────────────────────────────────────────

/**
 * Deliberately the same anatomy as [PosterCard], the venue card it sits beside
 * in the featured shelf: 150dp wide, a 168dp photo, then the title on the card's
 * own surface underneath. Events and venues are answers to the same question in
 * that row, so they must not look like two different kinds of object.
 *
 * Keep the two in step — a change to one card's frame, radius or type scale
 * belongs in both.
 */
@Composable
fun EventCard(event: FeedEvent, tonight: String, tomorrow: String, onOpen: () -> Unit) {
    Column(
        Modifier
            .width(150.dp)
            .clip(RoundedCornerShape(12.dp))
            .background(Theme.surface)
            .clickableUnlessBusy(onClick = onOpen),
    ) {
        Box(Modifier.fillMaxWidth().height(168.dp)) {
            EventPhoto(event.image, 168.dp, FeedImage.thumbWidthDp)
            Box(
                Modifier.fillMaxSize().background(
                    Brush.verticalGradient(
                        0.5f to Color.Transparent,
                        1.0f to Color.Black.copy(alpha = 0.55f),
                    ),
                ),
            )
            Box(Modifier.align(Alignment.TopStart).padding(6.dp)) { EventTag(event) }
        }

        Column(
            Modifier.padding(start = 10.dp, end = 10.dp, top = 8.dp, bottom = 10.dp),
            verticalArrangement = Arrangement.spacedBy(2.dp),
        ) {
            // Two lines on a reserved height, exactly as the venue card does it,
            // so every card in the row is the same height whatever the title.
            Text(
                event.displayTitle,
                fontFamily = Geist, fontWeight = FontWeight.SemiBold, fontSize = 13.sp,
                color = Theme.ink,
                maxLines = 2, overflow = TextOverflow.Ellipsis,
                modifier = Modifier.height(34.dp),
            )
            // Day and place only. The full meta line carries door times too,
            // which truncate away at this width — they are on the event page,
            // one tap down.
            Text(
                listOfNotNull(event.dayLabel(tonight, tomorrow), event.placeLine)
                    .joinToString(" · "),
                fontFamily = Geist, fontSize = 10.sp, color = Theme.fadedSand,
                maxLines = 1, overflow = TextOverflow.Ellipsis,
            )
        }
    }
}
